src/api/index.ts

```diff
@@ -0,0 +1,157 @@
+import createTilesApi, { TilesApi } from './tiles'
+import createTilesetsApi, { TilesetsApi } from './tilesets'
+
+export interface MapServerOptions {
+  dbPath: string
+}
+
+export interface Context {
+  activeImports: Map<string, MessagePort>
+  db: DatabaseInstance
+  piscina: Piscina
+  upstreamRequestsManager: UpstreamRequestsManager
+}
+
+// Any resource returned by the API will always have an `id` property
+export interface IdResource {
+  id: string
+}
+
+export interface Api
+  extends ImportsApi,
+    SpritesApi,
+    StylesApi,
+    TilesApi,
+    TilesetsApi {}
+
+function createApi(context: Context): Api {
+  const tilesetsApi = createTilesetsApi({ context })
+
+  const stylesApi = createStylesApi({
+    api: {
+      createTileset: tilesetsApi.createTileset,
+    },
+    context,
+  })
+
+  const importsApi = createImportsApi({
+    api: {
+      createTileset: tilesetsApi.createTileset,
+      createStyleForTileset: stylesApi.createStyleForTileset,
+    },
+    context,
+  })
+
+  const tilesApi = createTilesApi({
+    api: {
+      getTilesetInfo: tilesetsApi.getTilesetInfo,
+    },
+    context,
+  })
+
+  const spritesApi = createSpritesApi({
+    context,
+  })
+
+  return {
+    ...importsApi,
+    ...spritesApi,
+    ...stylesApi,
+    ...tilesApi,
+    ...tilesetsApi,
+  }
+}
+
+function init(dbPath: string): Context {
+  const db = new Database(dbPath)
+
+  // Enable auto-vacuum by setting it to incremental mode
+  // This has to be set before the anything on the db instance is called!
+  // https://www.sqlite.org/pragma.html#pragma_auto_vacuum
+  // https://www.sqlite.org/pragma.html#pragma_incremental_vacuum
+  db.pragma('auto_vacuum = INCREMENTAL')
+
+  // Enable WAL for potential performance benefits
+  // https://github.com/JoshuaWise/better-sqlite3/blob/master/docs/performance.md
+  db.pragma('journal_mode = WAL')
+
+  migrate(db, path.resolve(__dirname, '../../prisma/migrations'))
+
+  // Any import with an `active` state on startup most likely failed due to the server process stopping
+  // so we update these import records to have an error state
+  convertActiveImportsToErrorImports(db)
+
+  const piscina = new Piscina({
+    filename: path.resolve(__dirname, '../lib/mbtiles_import_worker.js'),
+    minThreads: 1,
+  })
+
+  piscina.on('error', (error) => {
+    // TODO: Do something with this error https://github.com/piscinajs/piscina#event-error
+    console.error(error)
+  })
+
+  return {
+    activeImports: new Map(),
+    db,
+    piscina,
+    upstreamRequestsManager: new UpstreamRequestsManager(),
+  }
+}
+
+const ApiPlugin: FastifyPluginAsync<MapServerOptions> = async (
+  fastify,
+  { dbPath }
+) => {
+  if (dbPath == null)
+    throw new Error('Map server option `dbPath` must be specified')
+
+  // Create context once for each fastify instance
+  const context = init(dbPath)
+
+  const api = createApi(context)
+
+  if (!fastify.hasDecorator('api')) {
```

Comment:
I'm unclear about when this might be true? I'm concerned that if it is true, then we might close the wrong instance of db and piscina below? Could the `onClose` hook also reference `this.api.close()` or something? And we could move the context & api creation into this if statement?

Reply:
Tbh not entirely sure - I plucked this from [this example](https://github.com/fastify/fastify-leveldb/blob/46c1f670580214242d1e6e60c8c2c8a234a34856/index.js#L31). My guess is that if the some other plugin decorated the instance using the same namespace?

> Could the onClose hook also reference this.api.close() or something?

Not entirely sure what you mean here. Does the api itslef need to some kind of closing functionality in that case? Since closing and the decorator are at the instance level, this seems to be okay as is to me

Reply:
Right, that makes sense (avoid namespace collision). If that's the case no need for api.close() (was only needed if context and api were not in scope, but in the if statement)
Good to merge I think